include/boost/url/string.hpp

```diff
@@ -20,33 +20,83 @@
 namespace boost {
 namespace urls {
 
-/** A read-only, reference counted string
+/** A read-only string
 
     Objects of this type represent read-only
     strings with ownership of the character
-    buffer. They are reference counted which
-    makes copies cheap. The type is derived
-    from @ref string_view which provides
-    compatibility with strings in terms of
-    comparisons and converisons. However,
-    care must be exercised; undefined
+    buffer.
+
+    They are reference counted to
+    makes copies more efficient.
+
+    The type is derived from `string_view`,
+    which provides compatibility with strings
+    in terms of comparisons and conversions.
+
+    However, care must be exercised; undefined
     behavior results if the string_view
     portion of the object is modified
     directly, for example by calling
     `remove_suffix` or `operator=`.
 
-    Slicing however, is supported, as
+    Slicing, however, is supported, as
     copies of the `string_view` portion
     of the object are valid and remain
-    valid for the lifetime of the oriignal
+    valid for the lifetime of the original
     object.
 */
 class string_value : public string_view
 {
+    /** Type for string_value reference counter and custom destructor
+
+        This struct stores the number of references to this read-only string.
+        As the string is read-only, we can create a copy by simply updating
+        the reference counter.
+
+        As the underlying string is constructed with a custom allocator, the
+        struct also includes a custom destructor to deallocate the
+        string whenever the reference counter reaches zero.
+
+     */
     struct base;
```

Comment:
Did you document a private nested type?

Reply:
We should never document private stuff, because the return on the investment is vanishingly small. First of all maintainers have a much better grasp of the library and its implementation than users. Second, users get zero value from documentation on private members. However, we should still use comments to explain non-obvious things in the implementation (inside function definitions). Almost always these comments should state "why" something is being done. Here's an example, we skip incrementing `it`:
https://github.com/vinniefalco/http_proto/blob/f2c3a5d9bfb49060af467b65843e125e6db958e2/include/boost/http_proto/rfc/impl/field_rule.ipp#L152

Reply:
> We should never document private stuff, because the return on the investment is vanishingly small

I wouldn't document everything that is private. I agree with the argument for not writing it. In this case, and only in this case, the marginal cost is zero because it was already done. So the return on investment in `NaN`.

> However, we should still use comments to explain non-obvious things in the implementation (inside function definitions). Almost always these comments should state "why" something is being done.

I can't disagree with that.

Reply:
keep the file